### PR TITLE
fix: some fixes for karpenter deploy

### PR DESCRIPTION
**Pull request by kvvit**
# Fixes for karpenter deploy

## First error:

```
Error: karpenter/private failed to create kubernetes rest client for update of resource: resource [karpenter.sh/v1/EC2NodeClass] isn't valid for cluster, check the APIVersion and Kind fields are valid

   with kubectl_manifest.ec2nodeclass_private[0],
   on main.tf line 71, in resource "kubectl_manifest" "ec2nodeclass_private":
   71: resource "kubectl_manifest" "ec2nodeclass_private" {
```

Solution: set `apiVersion: karpenter.k8s.aws/v1` for resources `kubectl_manifest`: `ec2nodeclass_private` and `ec2nodeclass_public` (file `terraform/modules/k8s-karpenter/main.tf`).

## Second error:

```
 Error: karpenter/default failed to run apply: error when creating "/tmp/879917200kubectl_manifest.yaml": NodePool.karpenter.sh "default" is invalid: [spec.disruption.consolidateAfter: Required value, spec.template.spec.nodeClassRef.group: Required value, spec.template.spec.nodeClassRef.kind: Required value, <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]

   with kubectl_manifest.nodepool["default"],
   on main.tf line 135, in resource "kubectl_manifest" "nodepool":
  135: resource "kubectl_manifest" "nodepool" {
```

Solution: Added fielsds `group` and `kind` for private and public NodePools (file `terragrunt/ACCOUNT_ID/us-east-1/demo/env.yaml`)

```yaml
        spec:
          nodeClassRef:
            group: karpenter.k8s.aws
            kind: EC2NodeClass
            name: public
```

## Third error:

```
 Error: karpenter/ci failed to run apply: error when creating "/tmp/969255726kubectl_manifest.yaml": NodePool.karpenter.sh "ci" is invalid: [spec.disruption.consolidateAfter: Required value, <nil>: Invalid value: "null": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]

   with kubectl_manifest.nodepool["ci"],
   on main.tf line 135, in resource "kubectl_manifest" "nodepool":
  135: resource "kubectl_manifest" "nodepool" {
```

Solution: Added field `consolidateAfter: 1m` for private and public NodePools (file `terragrunt/ACCOUNT_ID/us-east-1/demo/env.yaml`)

```yaml
     disruption:
       consolidationPolicy: WhenEmptyOrUnderutilized
       consolidateAfter: 1m
```

## Last one:

I removed the unnecessary data block `data "aws_ecrpublic_authorization_token" "token" {}` and the fields from the `helm_release` resource: `repository_username = data.aws_ecrpublic_authorization_token.token.user_name` and `repository_password = data.aws_ecrpublic_authorization_token.token.password`. I tested these changes, and everything works perfectly.

I also updated the `locals` block for Karpenter. I set the `chart` value to `oci://public.ecr.aws/karpenter`/karpenter and left the `repository` field empty. 

```
    chart         = try(var.helm.chart_name, "oci://public.ecr.aws/karpenter/karpenter")
    repository    = try(var.helm.repository, "")
```

This configuration has been tested across different Helm releases using OCI repositories, and everything works flawlessly. I successfully tested the Karpenter deployment from a Docker container in our sandbox account, and it was deployed without any issues.

There was another issue related to the `aws_ecrpublic_authorization_token`, related to aws provider. Since the public ECR is located in the `us-east-1` region, this data doesn't work in other regions. An additional AWS provider with an alias is required to handle this.

### CODE REVIEW

File: terraform/modules/k8s-karpenter/main.tf [hunk 1]
```diff
@@ -19,6 +19,12 @@ serviceAccount:
   annotations:
     eks.amazonaws.com/role-arn: ${module.this[0].iam_role_arn}
 
+postInstallHook:
```
Comment (mglotov) [inline terraform/modules/k8s-karpenter/main.tf:22]: Looks like the latest helm chart version doesn't have these values. Let's try to use the latest one.

File: terraform/modules/k8s-karpenter/main.tf [hunk 2]
```diff
@@ -3,8 +3,8 @@ locals {
   karpenter = {
     name          = try(var.helm.release_name, "karpenter")
     enabled       = true
-    chart         = try(var.helm.chart_name, "karpenter")
-    repository    = try(var.helm.repository, "oci://public.ecr.aws/karpenter")
+    chart         = try(var.helm.chart_name, "oci://public.ecr.aws/karpenter/karpenter")
```
Comment (mglotov) [inline terraform/modules/k8s-karpenter/main.tf:6]: Why did you change this? What was the problem here?

File: terraform/modules/k8s-karpenter/main.tf [hunk 3]
```diff
@@ -144,8 +148,6 @@ resource "helm_release" "this" {
   version             = local.karpenter.chart_version
   namespace           = module.namespace[count.index].name
   max_history         = 3
-  repository_username = data.aws_ecrpublic_authorization_token.token.user_name
```
Comment (mglotov) [inline terraform/modules/k8s-karpenter/main.tf:147]: I'm not sure it works. Let's test it together

File: terraform/modules/k8s-karpenter/main.tf (hunk 1 above)
Comment (kvvit) [reply terraform/modules/k8s-karpenter/main.tf]: Changed helm chart version to `1.0.3` and remoed `postInstallHook`

File: terraform/modules/k8s-karpenter/main.tf (hunk 2 above)
Comment (kvvit) [reply terraform/modules/k8s-karpenter/main.tf:6]: This configuration worked for github-runners that also use oci helm repository. And it's working now for deploy karpenter in Propeller project.

File: terraform/modules/k8s-karpenter/main.tf (hunk 3 above)
Comment (kvvit) [reply terraform/modules/k8s-karpenter/main.tf:147]: I tested this yesterday in the sandbox account, and in the  Propeller account. All works fine. More over, this data works only for `us-east-1` region. For other regions we need to use additional aws provider with alias.

File: terraform/modules/k8s-karpenter/main.tf (hunk 2 above)
Comment (kvvit) [reply terraform/modules/k8s-karpenter/main.tf:6]: I've added information about this changes in the description of the PR

**Review by mglotov — APPROVED** [review]

**Review by khalezin — APPROVED** [review]
